### PR TITLE
feat: add url_components for start/end

### PR DESCRIPTION
**Pull request by anonrig**

### CODE REVIEW

**Comment by lemire** [comment]
Let me try to do a PR on top to express my ideas.

**Comment by lemire** [comment]
Please see https://github.com/ada-url/ada/pull/251

**Comment by lemire** [comment]
With such a component, it could be fruitful to restructure the main ada/url. It could be made of a single string (which might possibly be identical to the input) along with this url_component. It could make the main ada/url structure smaller and faster. We could then boost our performance (e.g., by 25% or so).

**Review by lemire — APPROVED** [review]
The validation might be insufficient (???), but I like this PR a lot. If we can make sure that it is correct, it is a great step forward. 

Note that it is not documented in the README (and I think that's good at this time to be quiet about it).

**Comment by anonrig** [comment]
We also do not have all the tests for this. Let's finish them before merging this.

**Comment by lemire** [comment]
It would be good to have a plan to see where this is going. I started a discussion issue: https://github.com/ada-url/ada/discussions/253

**Comment by anonrig** [comment]
We are using a combination of `0` or `omitted` for the default values. We need to make sure the default values are properly tested too...

File: include/ada/url-inl.h [hunk 1]
```diff
@@ -73,6 +79,107 @@ namespace ada {
...
+    if (!path.empty()) {
+      out.pathname_start = uint32_t(trailing_index);
+      trailing_index += path.size();
+    } else {
+      out.pathname_start = uint32_t(trailing_index);
+    }
+
+    if (query.has_value()) {
+      out.search_start = uint32_t(trailing_index);
+      trailing_index += get_search().size();
```
Comment (lemire) [inline include/ada/url-inl.h:172]: trailing_index is fine. Another name for the same concept might be 'running_index'.

**Pull request merged**